File: modules/proximity/ios/ProximityModule.swift
import ExpoModulesCore

public class ProximityModule: Module {
  public func definition() -> ModuleDefinition {
    Name("Proximity")

    Events("onProximityChange")

    Function("setProximityMonitoringEnabled") { (enabled: Bool) in
      DispatchQueue.main.async {
        UIDevice.current.isProximityMonitoringEnabled = enabled
        
        if enabled {
          NotificationCenter.default.addObserver(
            self,
            selector: #selector(self.proximityStateDidChange),
            name: UIDevice.proximityStateDidChangeNotification,
            object: nil
          )
        } else {
          NotificationCenter.default.removeObserver(
            self,
            name: UIDevice.proximityStateDidChangeNotification,
            object: nil
          )
        }
      }
    }
  }

  @objc private func proximityStateDidChange() {
    let isNear = UIDevice.current.proximityState
    sendEvent("onProximityChange", [
      "isNear": isNear
    ])
  }
}

File: modules/proximity/src/ProximityModule.ts
import { NativeModule, requireNativeModule } from 'expo';

declare class ProximityModule extends NativeModule<{
  onProximityChange: (event: { isNear: boolean }) => void;
}> {
  setProximityMonitoringEnabled(enabled: boolean): void;
}

export default requireNativeModule<ProximityModule>('Proximity');
